src/main/java/org/broadinstitute/hellbender/tools/reference/ReferenceSequenceTable.java

```diff
@@ -0,0 +1,354 @@
+                int length = record.getSequenceLength();
+                String md5 = calculateMD5(record, entry.getValue());
+                GATKPath reference = entry.getKey();
+                TableEntry newEntry = new TableEntry(getReferenceDisplayName(reference), name);
+                TableRow newRow = null;
+
+                // map each MD5 to List of TableEntry objects containing length, md5, and name
+                if (!tableByMD5.containsKey(md5)) {
+                    newRow = new TableRow(md5, length, new ArrayList<>(referenceSources.keySet()));
+                    tableByMD5.put(md5, newRow);
+                }
+                tableByMD5.get(md5).add(newEntry);
+
+                if(!tableBySequenceName.containsKey(name)){
+                    Set<TableRow> rowSet = new HashSet<>();
+                    tableBySequenceName.put(name, rowSet);
+                }
+                tableBySequenceName.get(name).add(tableByMD5.get(md5));
+            }
+        }
+    }
+
+    public Set<String> getAllSequenceNames(){
+        return tableBySequenceName.keySet();
+    }
+
+    /**
+     * Given an MD5, returns its corresponding row
+     *
+     * @param md5 The MD5 as a String
+     * @return the corresponding TableRow from the tableByMD5
+     */
+    public TableRow queryByMD5(String md5){
+        return tableByMD5.get(md5);
+    }
+
+    /**
+     * Given a sequence name, returns the set of its corresponding rows
+     *
+     * @param sequenceName The sequence name as a String
+     * @return the set of TableRows that contain the sequence name
+     */
+    public Set<TableRow> queryBySequenceName(String sequenceName){
+        return tableBySequenceName.get(sequenceName) == null ? Collections.emptySet() : tableBySequenceName.get(sequenceName);
+    }
+
+    private String calculateMD5(SAMSequenceRecord record, ReferenceDataSource source) {
+        final String md5FromDict = record.getMd5();
+        final SimpleInterval referenceInterval = new SimpleInterval(record.getSequenceName(), 1, record.getSequenceLength());
+        String md5 = null;
+
+        switch (md5Mode) {
+            case USE_DICT:
+                if (md5FromDict == null || md5FromDict.isEmpty()) {
+                    throw new UserException.BadInput("Running in USE_DICT mode, but MD5 missing for sequence " + record.getSequenceName() + ". Run --md5-calculation-mode with a different mode to recalculate MD5.");
+                }
+                md5 = md5FromDict;
+                break;
+            case RECALCULATE_IF_MISSING:
+                if (md5FromDict == null || md5FromDict.isEmpty()) {
+                    md5 = ReferenceUtils.calculateMD5(source, referenceInterval);
+                }
+                else {
+                    md5 = md5FromDict;
+                }
+                break;
+            case ALWAYS_RECALCULATE:
+                md5 = ReferenceUtils.calculateMD5(source, referenceInterval);
```

Comment:
There should be a check here if this md5 mismatches with the md5 from the dictionary (if it exists of course). If it ever does we should use the logger to throw a warning to user listing the two md5s and the contig name and the reference for which it is mismatching.